proposals/0000-specialise-expressions.rst

```diff
@@ -0,0 +1,248 @@
+  {-# RULES "addMult/Int" addMult = addMultInt #-}  -- NB: applies only at type Int
+
+Note that ``addMultInt`` uses no indirection in its use of addition and multiplication,
+unlike ``addMult``, which makes two calls to dictionary functions, confounding branch
+prediction and slowing down execution. The ``RULE`` means that GHC will replace
+all calls of ``addMult`` at ``Int`` to use the faster version.
+
+The syntax of today's ``SPECIALISE`` requires a function and its specialised type.
+
+This proposal suggests to generalise the syntax of ``SPECIALISE`` to allow an
+arbitrary expression, not just a function name and a type. Happily, this is backward
+compatible, because type signatures are expressions, too. This will allow, for example ::
+
+  {-# SPECIALISE addMult @Double #-}
+  {-# SPECIALISE addMult (5 :: Int) #-}
+
+The first of these is a more succinct phrasing of what we have today: it simply
+specialises ``addMult`` to work on ``Double``\ s, but without having to repeat the
+entire type. The second does something new: it allows expression specialisation,
+which can then lead to further evaluation within the function body. In this case,
+GHC will create ::
+
+  addMult5 :: Int -> Int -> Int
+  addMult5 y = let x = 5 in x * 2 + y
+  {-# RULES "addMult/5" addMult 5 = addMult5 #-}
+
+GHC can then further optimize the right-hand side of ``addMult5`` to avoid doing
+any multiplication at run time.
+
+Motivation
+----------
+See the introduction. In addition to the examples there, we can imagine
+a case like this ::
+
+  frob :: Bool -> ...
+  frob b ... = ...
+    where
+      ... if b then helper1 else helper2 ...
+
+The function ``frob`` is a big, complicated function. Deep in its bowels, it
+either uses ``helper1`` or ``helper2`` to do a critical step. However, most (or maybe all)
+call sites of ``frob`` pass in a literal ``True`` or ``False``. Really, it would
+make sense to have ``frobHelper1`` and ``frobHelper2`` as separate functions, but
+that would cause unhappy duplication of all the complexity inherent in ``frob``.
+
+With the ability to specialise expressions, we can write ::
+
+  {-# SPECIALISE frob True #-}
+  {-# SPECIALISE frob False #-}
+
+This will make two copies of ``frob``, one for ``True`` and one for ``False``. These
+will then be optimised to make direct calls to ``helper1`` or ``helper2``, respectively.
+Call sites (that use a literal ``True`` or ``False``) will be rewritten to use the
+specialised versions.
+
+Proposed Change Specification
+-----------------------------
+
+1. Here is the new BNF for ``SPECIALISE`` pragmas::
+
+     pragma ::= ...
+             |  '{-#' specialise_keyword activation rule_foralls specexp '#-}'
+             |  '{-#' specialise_keyword activation qvar '::' types1 '#-}'  -- as today
+
+     specialise_keyword ::= 'SPECIALISE' | 'SPECIALIZE' | 'SPECIALISE INLINE' | 'SPECIALISE INLINE'
+
+     specexp ::= qvar
+              |  specexp aexp
+              |  specexp '@' atype
+
```

Comment:
This would allow `{-# SPECIALISE addMult 5 :: Int -> Int #-}`, although it makes `{-# SPECIALISE qvar :: type #-}` ambiguous. (Perhaps we can resolve the ambiguity by having the "backwards compatibility alternative" kick in only when there are at least two types?)

```suggestion
              |  specexp :: type
              
```

Reply:
I have fixed this